Bump `dev-haskell/hmatrix`

waivio opened the pull request.

waivio commented:
Let me know if I need to do something else.
Thanks,
Nathan.

hololeap commented on dev-haskell/hmatrix/hmatrix-0.20.2.ebuild:
```diff
@@ -0,0 +1,40 @@
+	dev-haskell/storable-complex:=[profile?]
+	>=dev-haskell/vector-0.11:=[profile?]
+	>=dev-lang/ghc-8.4.3:=
+	openblas? ( unknown-c-lib/openblas )
```
This will need to be changed to:

```
openblas? ( sci-libs/openblas )
```

waivio replied:
changed as requested

hololeap commented on dev-haskell/hmatrix/Manifest:
```diff
@@ -1 +1,2 @@
 DIST hmatrix-0.20.0.0.tar.gz 102180 BLAKE2B fbbe38a89813fc4ca8fdfc21a46cd5ae36a5e0fe05f742aceb2febdfbc6a2fcd1653bc5887859d6d3ff49841a1265253765cee4679c8f6338978a808a7bec18d SHA512 1845bd5a88d589a7b7543de3543747a3de642ca0801e1483e5022dd34cffc2229b478d6f91fde5fe0d525365c8644ce4a67da9bc32ea2872845531f339947d7c
```
You might as well delete the old version since it's no longer building.

waivio replied:
removed old version and patches, cleaned up the manifest as requested.

hololeap replied:
Thanks, although it looks like you need to use `git mergetool` or manually get rid of everything execpt line 4

hololeap reviewed (changes requested):
Here are some changes that will need to be made. Have you tried building this?

hololeap reviewed (changes requested):
Please change the commit message to something like:

```
dev-haskell/hmatrix: Bump to 0.20.2
```

hololeap commented:
@waivio Thanks for taking charge on this. Here are just some general tips to help you meet our QA standards:

- [ ] You can take this opportunity to get familiar with the [`repoman`](https://wiki.gentoo.org/wiki/Repoman) tool. Running `repoman -dx fix` in the package directory (`dev-haskell/hmatrix/`) will check for QA errors and fix the `Manifest` for you.
- [ ] `repoman` [can also commit for you](https://wiki.gentoo.org/wiki/Repoman#Invocation) which will open a text editor and automatically add the package name to the start of the commit message.
- [ ] You should also do a bit of testing by trying to build the package on your system and possibly try out different combinations of `USE` flags.
- [ ] It may be a good idea to check if related packages will build with the version bump, or if they need a version bump themselves. In this case, these packages should be tested at a minimum:
    - [ ] `dev-haskell/hmatrix-gsl-stats`
    - [ ] `dev-haskell/hmatrix-gsl`
    - [ ] `dev-haskell/hmatrix-vector-sized`
- [ ] You can also [check for reverse dependencies](https://github.com/gentoo-haskell/gentoo-haskell/wiki/Useful-snippets#find-reverse-dependencies-of-a-package) and test them. In this case, these packages are reverse dependencies. (I can check these for you if you don't get to it):
    - [ ] `dev-haskell/hamilton`
    - [ ] `dev-haskell/hstatistics`
    - [ ] `dev-haskell/picedit`
    - [ ] `dev-haskell/sibe`

hololeap commented:
I turned that last comment into a wiki article:
https://github.com/gentoo-haskell/gentoo-haskell/wiki/General-QA-tips

solpeth commented:
@hololeap could you add your wiki article to the relevant doc in the repo on contributing tips? I can't remember which file it is exactly but it would go nicely with the other bumping things to look out for (like revbump tips).

Also I'm afk but I have a sneaking suspicion that there's a Gentoo openblas file in FILESDIR that needs to be added to the ebuild. Could be completely wrong as just going off memory here. Could be some other openblas user.

waivio commented:
Squashed commits and changed commit message as requested.
Ran repoman and it appeared to be happy with the changes and it updated the manifest file.
How do I test the changes?
Do I have to somehow make a local overlay and emerge from that?

waivio commented:
Apparently my git work corrupted the manifest file somehow.  I'll have to try and fix.

hololeap commented:
@waivio there is more than one way to do this.

1. You can point the `[haskell]` entry in `/etc/portage/repos.conf/*` to the repo copy you are working on:
   ```
   [haskell]
   location = /home/user/gentoo-haskell
   sync-type = git
   sync-uri = https://github.com/gentoo-haskell/gentoo-haskell
   auto-sync = No
   ```
   (You will have to manually sync and make sure it's on the correct branch)
2. You can add a new entry which overrides the existing one:
   ```
   [haskell-waivio]
   location = /var/db/repos/haskell-waivio
   sync-type = git
   sync-uri = https://github.com/waivio/gentoo-haskell
   sync-git-clone-extra-opts = --branch bump-hmatrix-to-0.20.2
   priority = -10
   ```

(I haven't tested # 2)

waivio commented:
When I re-ran repoman it now has complaints about invalid IUSE, it looks like the 0.20 version removed all of the IUSE and the patch edited the cabal file to delete the openblas flag logic for different operating systems.  Does Gentoo use virtual/blas and virtual/lapack to switch between blas implementations?  I'm thinking I need to reproduce what was done in the version 0.20.0.0 ebuild and add the patch back and modify it somehow.

hololeap commented:
Hm, `hackport` should have added `<use>` entries to `metadata.xml`. Did you forget to add it before committing?

hololeap commented on dev-haskell/hmatrix/hmatrix-0.20.0.0.ebuild:
```diff
@@ -1,33 +0,0 @@
-	>=dev-haskell/cabal-1.24.0.0
-"
-
-PATCHES=( "${FILESDIR}"/${PN}-0.20.0.0-gentoo-blas.patch )
```
It looks like the previous author of the ebuild used this patch to remove `openblas` support entirely. You can try adding this line (and the associated patch) back in. You will also need to remove the `openblas` USE flag and the associated entry in `metadata.xml`.

If the patch fails to apply when you try building it, you can create a new patch for this version. (I can share my workflow in this regard if we get there.)

hololeap commented:
Btw, @waivio , if you want to chat you can join `#gentoo-haskell` in [libera.chat](https://libera.chat/guides/webchat)

waivio commented:
Changes:
* Added ebuild for hmatrix 0.20.2 via hackport with some manual edits
* Updated Manifest via repoman
* Updated metadata.xml to the latest description in Hackage
* Added new patch for 0.20.2 via diff

I plan on testing it later.

It appears, after investigation, that the patch is to compensate for the differences in how multiple implementations of blas/lapack are handled for varying operating systems.  Gentoo uses eselect to choose the blas/lapack implementation providing a virtual layer for cabal to link-to rather than needing to specify different paths for the implementations in cabal.

hololeap commented on dev-haskell/hmatrix/hmatrix-0.20.2.ebuild:
```diff
@@ -0,0 +1,34 @@
+LICENSE="BSD"
+SLOT="0/${PV}"
+KEYWORDS="~amd64 ~x86"
+IUSE=""
```
The .cabal file still contains two flags after your patch is applied: `disable-default-paths` and `no-random_r`. If you want to have these permanently off, you will want to add a `src_configure` block that reflects this:

```
src_configure() {
	haskell-cabal_src_configure \
		--flag=-disable-default-paths \
		--flag=-no-random_r
}
```

You should also put this line underneath `# ebuild generated by hackport ...`, which will tell hackport to disable these in future bumps to the package:

```
#hackport: flags: -disable-default-paths,-no-random_r,-openblas
```

hololeap replied:
Also, you can remove `IUSE=""` if it is empty

hololeap replied:
You can also add these back in if you want:

```
#hackport: flags: no-random_r:no-random-r,-openblas
```

...

```
IUSE="disable-default-paths no-random-r"
```

...

```
src_configure() {
	haskell-cabal_src_configure \
		$(cabal_flag disable-default-paths disable-default-paths) \
		$(cabal_flag no-random-r no-random_r)
}
```

`metadata.xml`
```
	<use>
		<flag name="disable-default-paths">When enabled, don't add default hardcoded include/link dirs by default. Needed for hermetic builds like in nix.</flag>
		<flag name="no-random-r">When enabled, don't depend on the random_r() C function.</flag>
	</use>
```

hololeap replied:
Something that can be helpful is generating an ebuild directly from the modified cabal file you were working on. For instance:

```bash
hackport make-ebuild dev-haskell "somepath/hmatrix-r1/hmatrix.cabal"
```

This will adjust the modified dependencies for you and make the whole process less cumbersome.

waivio replied:
The `hackport make-ebuild` technique was definitely the way to go.
`dos2unix` didn't appear to make any changes.

hololeap commented on dev-haskell/hmatrix/files/hmatrix-0.20.2.0-gentoo-blas.patch:
```diff
@@ -0,0 +1,70 @@
+diff -urN hmatrix/packages/base/hmatrix.cabal hmatrix-r1/packages/base/hmatrix.cabal
+--- hmatrix/packages/base/hmatrix.cabal	2022-02-27 21:37:00.805021893 -0700
++++ hmatrix-r1/packages/base/hmatrix.cabal	2022-02-27 21:43:12.529392192 -0700
```
There's something off about the paths here that's causing `patch` to fail. However, it works if I adjust it to this:

```diff
diff -urN hmatrix/hmatrix.cabal hmatrix-r1/hmatrix.cabal
--- hmatrix/hmatrix.cabal	2022-02-27 21:37:00.805021893 -0700
+++ hmatrix-r1/hmatrix.cabal	2022-02-27 21:43:12.529392192 -0700
```

waivio replied:
Rebuilt the patch.  I must have needed to be 1 directory level down from the cabal file.

waivio commented:
@hololeap , Yes it looks like the previous bumps may have been perpetuating the status quo for expediency. 

I did a brief investigation into the Cabal flags for 'hmatrix'.

* 'hmatrix' PR#301 for the 'no-random-r' flag states:

> Allow disabling random_r() usage manually. See PR279.
> This allows building hmatrix against the musl libc, which allows
> easily linking Haskell programs statically. A feature-detection for
> random_r() would be even better, but this will do for now.

'musl' support sound good to me, so I will try and add that flag back.  The upstream followup issue, in the quote, has not been implemented.  Is there a better flag to indicate needed for 'musl' support?  I couldn't find any.


* 'hmatrix' PR#219 for the 'disable-default-paths' the original committer stated:

> It is convenient for the users to be able to just install it without
> having to know / learn about the own distribution's paths -- as a nice
> service kindly provided by the package maintainer. It will eventually 
> become unnecessary with the upstream project (openblas) providing
> pkgconfig. But until then I find it is an acceptably clean solution to
> have this disable-default-paths flag that disables these defaults.

I believe that Gentoo supports pkg-config and/or the 'eselect' method for 'openblas', so I'm thinking 'disable-default-paths' probably not necessary.

For the 'openblas' flag itself I thing the 'eselect' tool can take care of everything.

What are your thoughts?

hololeap commented:
Honestly, that all sounds pretty reasonable, but the fact is I don't understand much of this stuff (`blas` vs `openblas`, `glibc` vs `musl`) well enough to give any good advice. My general advice would be get this working "well enough" and if anyone needs those other features, we can look at adding them in later. However, if you want to try to get this stuff working, you are more than welcome to.

The previous author, @trofi, has retired, but they may be willing to chime in here. You might want to ask in the `#gentoo-dev-help` channel on `libera.chat`.

trofi commented:
For `musl` support `.ebuild` switch mechanism used to be based on `ELIBC_musl`/`ELIBC_glibc` `USE_EXPAND`s. Maybe things have changed. Worth checking out `::gentoo`. I'm surprised bindings care about libc details at all.

As for blas implementation selection I don't think there is a good way to deal with it at build time in Gentoo. I has always been lazy and just pulled in `virtual/blas` to have it working. AFAIU it's not ideal as it's a reasonable setup to have multiple BLAS implementations in the system (but maybe not anymore?). https://www.mail-archive.com/gentoo-dev@lists.gentoo.org/msg80795.html proposed `blas.eclass` a while ago but it did not go through. Probably worth asking sci project (and other blas Gentoo maintainers) for today's guidelines on packaging `blas` revdeps.

hololeap commented:
I tested these packages and they all compile successfully after the bump to `dev-haskell/hmatrix`:

- [x] `dev-haskell/hmatrix-gsl-stats`
- [x] `dev-haskell/hmatrix-gsl`
- [x] `dev-haskell/hmatrix-vector-sized`
- [x] `dev-haskell/hamilton`
- [x] `dev-haskell/hstatistics`
- [x] `dev-haskell/picedit`
- [x] `dev-haskell/sibe`

So, once everyone is satisfied with the state of `dev-haskell/hmatrix`, I can go ahead and pull this.

waivio commented:
After reviewing [rework to stop altering library symlinks](https://bugs.gentoo.org/632624) :
> As discussed in the gentoo-dev, it is decided to adopt the runtime
> switching of BLAS/CBLAS that has been proved by the experience of
> Debian.  The Gentoo libgcc-style ld.so.conf runtime switching has
> been adopted as well, so as to avoid changing symlinks in /usr
 > ignoring package manager.

> This design solves all the remaining problems regarding BLAS/LAPACK,
> thus making alternative-2.eclass (runtime switching of build
> options) and proposed USE_EXPAND (PYTHON_TARGET like eclass)
 > obsolete.

> Finally we are at it.  Thanks to all for years of ideas, debates and
> hard work.

I'm not sure what that means...
But, I think you can change implementations of blas/lapack with eselect:

> $ eselect blas list
> Available BLAS/CBLAS (lib) candidates:
>   (none found)
> Available BLAS/CBLAS (lib64) candidates:
>   [1]   openblas
>   [2]   reference *

> $ eselect lapack list
> Available LAPACK (lib) candidates:
>   (none found)
> Available LAPACK (lib64) candidates:
>   [1]   openblas
>   [2]   reference *

hololeap commented:
Looks good. I'm just going to merge your two commits with `git rebase -i master`:

```
pick d7a5e333ff dev-haskell/hmatrix: Bump to 0.20.2
fixup 7ad520f599 Rebuilt to preserve the 'musl' compatibility flag, and updated the patch file
```

hololeap commented:
Pulled. Thanks!
